movement/lib/TOTP/sha1.c

```diff
@@ -0,0 +1,398 @@
+};
+
+/*
+ * SHA-1 final digest
+ */
+void mbedtls_sha1_finish( mbedtls_sha1_context *ctx, unsigned char output[SHA1_DIGEST_LENGTH] )
+{
+    uint32_t last, padn;
+    uint32_t high, low;
+    unsigned char msglen[8];
+
+    high = ( ctx->total[0] >> 29 )
+         | ( ctx->total[1] <<  3 );
+    low  = ( ctx->total[0] <<  3 );
+
+    PUT_UINT32_BE( high, msglen, 0 );
+    PUT_UINT32_BE( low,  msglen, 4 );
+
+    last = ctx->total[0] & 0x3F;
+    padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );
+
+    mbedtls_sha1_update( ctx, sha1_padding, padn );
+    mbedtls_sha1_update( ctx, msglen, 8 );
+
+    PUT_UINT32_BE( ctx->state[0], output,  0 );
+    PUT_UINT32_BE( ctx->state[1], output,  4 );
+    PUT_UINT32_BE( ctx->state[2], output,  8 );
+    PUT_UINT32_BE( ctx->state[3], output, 12 );
+    PUT_UINT32_BE( ctx->state[4], output, 16 );
+}
+
+/*
+ * output = SHA-1( input buffer )
+ */
+void mbedtls_sha1( const unsigned char *input, size_t ilen, unsigned char output[SHA1_DIGEST_LENGTH] )
+{
+    mbedtls_sha1_context ctx;
+
+    mbedtls_sha1_init( &ctx );
+    mbedtls_sha1_starts( &ctx );
+    mbedtls_sha1_update( &ctx, input, ilen );
+    mbedtls_sha1_finish( &ctx, output );
+    mbedtls_sha1_free( &ctx );
+}
+
+/*
+* Compute HMAC_SHA1 using key, key length, text to hash, size of the text, and output buffer
+*/
+void HMAC_SHA1(const uint8_t* key, size_t key_length, const uint8_t *in, size_t n, uint8_t out[SHA1_DIGEST_LENGTH]){
+
+  uint8_t i;
+  uint8_t k_ipad[SHA1_BLOCK_LENGTH]; /* inner padding - key XORd with ipad */
+  uint8_t k_opad[SHA1_BLOCK_LENGTH]; /* outer padding - key XORd with opad */
+  uint8_t buffer[SHA1_BLOCK_LENGTH + SHA1_DIGEST_LENGTH];
+
+  /* start out by storing key in pads */
+  memset(k_ipad, 0, sizeof(k_ipad));
+  memset(k_opad, 0, sizeof(k_opad));
+
+  if (key_length <= SHA1_BLOCK_LENGTH) {
+      memcpy(k_ipad, key, key_length);
+      memcpy(k_opad, key, key_length);
+  }
+
+  else {
+      mbedtls_sha1(key, key_length, k_ipad);
+      memcpy(k_opad, k_ipad, SHA1_BLOCK_LENGTH);
+  }
+
+  /* XOR key with ipad and opad values */
+  for (i = 0; i < SHA1_BLOCK_LENGTH; i++) {
+      k_ipad[i] ^= HMAC_IPAD;
+      k_opad[i] ^= HMAC_OPAD;
+  }
+  
+  // perform inner SHA1
+  memcpy(buffer, k_ipad, SHA1_BLOCK_LENGTH);
+  memcpy(buffer + SHA1_BLOCK_LENGTH, in, n);
+  mbedtls_sha1(buffer, SHA1_BLOCK_LENGTH + n, out);
+  
+  memset(buffer, 0, SHA1_BLOCK_LENGTH + n);
+
+  // perform outer SHA1
+  memcpy(buffer, k_opad, SHA1_BLOCK_LENGTH);
+  memcpy(buffer + SHA1_BLOCK_LENGTH, out, SHA1_DIGEST_LENGTH);
+  mbedtls_sha1(buffer, SHA1_BLOCK_LENGTH + SHA1_DIGEST_LENGTH, out);
+}
+/*
+* Compute TOTP_HMAC_SHA1 using key, key length, text to hash, size of the text
+*/
+uint32_t TOTP_HMAC_SHA1(const uint8_t* key, size_t key_length, const uint8_t *in, size_t n){
+    // STEP 1, get the HMAC-SHA1 hash from counter and key
+    uint8_t hash[SHA1_DIGEST_LENGTH];
+    HMAC_SHA1(key, key_length, in, 8, hash);

```

Comment:
It seems here like we're ignoring the text size and hard-coding a length of 8 in the call to `HMAC_SHA1`. Shouldn't this be:

```c
HMAC_SHA1(key, key_length, in, n, hash);
```

I see the same issue in `TOTP_HMAC_SHA256` and `TOTP_HMAC_SHA512`.

Reply:
You are right :). This was fixed in https://github.com/joeycastillo/Sensor-Watch/pull/151/commits/db2b82a5dc0e310489e966d1cff902568627d642.